### PR TITLE
Migrate NapiJsiRuntime from V8-JSI

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description

Moves the implementation of MakeNapiJsiRuntime from V8-JSI into React Native Windows.

## Motivation

MakeNapiJsiRuntime exposes non-ABI-safe C++ JSI APIs.
Thus, it should not reside in the V8-JSI DLL boundary.
As an intermediate measure, moving such method and supporting types into RNW.

## Pre-merge Checklist
- [x] Migrate integration test.
- [x] Address TODO comments.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/8617)

### CODE REVIEW

File: vnext/Shared/JSI/NapiJsiRuntime.h
Comment (tudorms) [inline vnext/Shared/JSI/NapiJsiRuntime.h:19]: >__cdecl [](http://example.com/codeflow?start=40&length=7)

(Nit) Probably no longer needed since we don't export it

**Review by tudorms — APPROVED** [review]
:shipit:

File: vnext/Shared/JSI/NapiJsiRuntime_detail.h
Comment (vmoroz) [inline vnext/Shared/JSI/NapiJsiRuntime_detail.h:1]: >/ Copyri [](http://example.com/codeflow?start=1&length=8)

Please move it back to the NapiJsiRuntime.cpp

**Review by vmoroz — CHANGES_REQUESTED** [review]
:clock1:

File: vnext/Shared/JSI/NapiJsiRuntime_detail.h
Comment (vmoroz) [inline vnext/Shared/JSI/NapiJsiRuntime_detail.h:14]: It is not correct C++

File: vnext/Shared/JSI/NapiJsiRuntime_detail.h
Comment (vmoroz) [inline vnext/Shared/JSI/NapiJsiRuntime_detail.h:19]: Anonymous namespaces must be never used in the header files. It produces a not-optimal code.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:4]: I would expect that code does not compile without it. I think we use some experimental functions.

**Comment by vmoroz** [comment]
Please make sure that we run the JSI tests for the new code.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:589]: all these CreateSymbol (and probably other property ID from name) will result in strings that will be duplicated across every single instance. Make the strings class statics const to avoid the string duplication since the utf8 string will always be the same and not memoized this way

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:615]: this seems wasteful (esp for large script buffers), can we do away with some of these str copies?

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1202]: typo, GetHostFunction

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1302]: should there be a special case for hresult_error?

Comment (asklar) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:1302]: same in the next function

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1697]: need to ensure T is not an array type since you'd have to call `delete[]`  instead

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1781]: is capturing by reference, safe here? if so pls add a comment explaining why

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:4]: It does compile, but I'll restore the define since you say it IS used.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:615]: Not sure. Are these copies necessary @tudorms, @vmoroz?

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:1302]: Can't hurt, but it's not clear from this call chain whether arbitrary, potentially winrt-throwing functions could be uses.

@vmoroz is `lambda` a potentially arbitrary function object that could throw any exception type?

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:1781]: I'm not aware of the threading model for these.
Note, I have made captures explicit, while still keeping them by reference.

@vmoroz are these captures thread-safe?

File: vnext/Shared/JSI/NapiJsiRuntime.h
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.h:19]: Added the symbol to the DLL boundary to make it accessible for integration tests.
Keeping the explicit calling convention for now.

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj
Comment (JunielKatarn) [inline vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj:33]: Remove before merging.

File: vnext/Desktop.ABITests/NapiTests.cpp
Comment (vmoroz) [inline vnext/Desktop.ABITests/NapiTests.cpp:1]: >// [](http://example.com/codeflow?start=0&length=2)

The file name should be JSI tests - it does not run NAPI tests - it runs JSI tests for the JSI implemented on top of the NAPI.

File: vnext/Desktop.DLL/react-native-win32.x64.def
Comment (vmoroz) [inline vnext/Desktop.DLL/react-native-win32.x64.def:66]: >MakeNapiJsiRuntime2 [](http://example.com/codeflow?start=1&length=19)

Ideally we must not export this function, but instead add the source files to the Nuget package for other project to compile them outside of the ABI boundary.

File: vnext/Shared/JSI/NapiJsiRuntime.h
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.h:19]: Since we do not want this function to be in the ABI boundary, it would be better if tests compile this code as the consumers of the DLL must do.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:1360]: Since we must distribute this code as source, it would be better to avoid any dependency on C++/WinRT for now. Otherwise, we will not be able to use it in Office.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: There will be no C++ constant string duplications, but the NapiRefHolder cannot be global, it must be per the V8 engine runtime instance.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: Yes, they are thread safe because the lamba is executed synchronously and never saved or passed to another thread.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: It is only used internally by the code in this .cpp file. We do not use T[] here.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:636]: @asklar, could you explain which string copy you see here are wasteful?

File: vnext/Desktop.DLL/react-native-win32.x64.def
Comment (JunielKatarn) [reply vnext/Desktop.DLL/react-native-win32.x64.def:66]: We will remove it eventually, but we don't have a test project that:
1. Uses GoogleTest.
2. Statically links the library that includes `NapiJsiRuntime.cpp`.

The test is implemented in an integration test project (meaning, it consumes the DLL and only has access to its exported symbols).

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:636]: 1. `CreateStringUtf8` will call into napi which will copy the string
2. `SerializeScript` calls `vector::assign` with the 2nd overload, which copies the string

The VectorBuffer ctor does a move so that one is ok.

File: vnext/Desktop.DLL/react-native-win32.x64.def
Comment (vmoroz) [reply vnext/Desktop.DLL/react-native-win32.x64.def:66]: Yes, these are new files and we need to modify an existing unit test project to do that. The main point is that the NapiJsiRuntime.h and NapiJsiRuntime.cpp must be compiled as a part of the project that uses the RN DLL.
The ideal place for these files to satisfy these needs in the Microsoft.ReactNative.Cxx project where we have other .h/.cpp files which we want to be compiled outside of the DLL to use the ABI-safe APIs.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1561]:
```suggestion
      "Cannot convert a non JS string NodeApi Value to a std::string.");
```

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1342]: I remember in Com days with SetLastError and GetLastError we had to 'backup' the previous error state sometimes.
From the name of themethod NodeApi errors are implemented similar to windows and do not contain a 'stack' of errors.

Is there a chance we are potentially losing a previous error? Especially since we are in error handling mode here?
I.e. should we get the last exception before we call `napi_get_property` and if present 'put the error back'?

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1345]: Can V8/NodeApi be localized. it seems 'risky' checking for an error substring here?

Comment (dannyvv) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:1345]: If not the message below, should probably be localized by/for RN/RNW

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.cpp:1359]: Could this obfuscate the callstack?

File: vnext/Shared/JSI/NapiJsiRuntime.h
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.h:16]:
```suggestion
// NodeApiJsiRuntime factory function.
```

**Comment by dannyvv** [comment]
@JunielKatarn: Since I'm talking about the rename to `NodeApi` Let me know if you'd like me to do it for you or not.

File: vnext/Shared/JSI/NapiJsiRuntime.h
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiRuntime.h:19]:
```suggestion
std::unique_ptr<facebook::jsi::Runtime> __cdecl MakeNodeApiJsiRuntime(napi_env env) noexcept;
```

File: vnext/Shared/JSI/NapiJsiV8RuntimeHolder.cpp
Comment (dannyvv) [inline vnext/Shared/JSI/NapiJsiV8RuntimeHolder.cpp:95]:
```suggestion
  m_runtime = MakeNodeApiJsiRuntime(env);
```

File: vnext/Desktop.DLL/react-native-win32.x64.def
Comment (dannyvv) [inline vnext/Desktop.DLL/react-native-win32.x64.def:66]:
```suggestion
?MakeNodeApiJsiRuntime@JSI@Microsoft@@YA?AV?$unique_ptr@VRuntime@jsi@facebook@@U?$default_delete@VRuntime@jsi@facebook@@@std@@@std@@PEAUnapi_env__@@@Z
```

File: vnext/Desktop.DLL/react-native-win32.x86.def
Comment (dannyvv) [inline vnext/Desktop.DLL/react-native-win32.x86.def:68]:
```suggestion
?MakeNodeApiJsiRuntime@JSI@Microsoft@@YA?AV?$unique_ptr@VRuntime@jsi@facebook@@U?$default_delete@VRuntime@jsi@facebook@@@std@@@std@@PAUnapi_env__@@@Z
```

**Review by dannyvv — APPROVED** [review]
Disclaimer, didn't go over every single line in full detail in JsiRuntime.cpp file

File: vnext/Desktop.DLL/react-native-win32.x64.def
Comment (JunielKatarn) [reply vnext/Desktop.DLL/react-native-win32.x64.def:66]: I'll look into both aspects after merging this case, but I don't know whether `Microsoft.ReactNative.Cxx` can be consumed by the Desktop DLL.

File: vnext/Desktop.ABITests/JsiTests.cpp
Comment (vmoroz) [inline vnext/Desktop.ABITests/JsiTests.cpp:30]: >MakeNapiJsiRuntime2 [](http://example.com/codeflow?start=31&length=19)

This is non-ABI safe API. The tests are suppose to test only the ABI-safe APIs.

Comment (JunielKatarn) [reply vnext/Desktop.ABITests/JsiTests.cpp:30]: This non-ABI safe API is both defined and used within the RNW DLL.
Is this a problem at all?

Comment (JunielKatarn) [reply vnext/Desktop.ABITests/JsiTests.cpp:30]: Thanks for the off-line clarification.

This methods is not meant to be exported, and for non-RNW projects consuming it, we will export the source files as part of the NuGet package.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:214]: This was suggested by @vmoroz to keep the implementation self-contained the code generation efficient (using anonymous namespaces in both header and source file may lead to redundant linkage).

A first iteration of this PR actually separated declarations into a different header, but it was reverted due to the technicality mentioned above.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:279]: Got the idea.

I will apply the renaming.

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp:653]: For functions, definitely. For types, we have made exceptions.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Shared/JSI/NapiJsiRuntime.cpp
Comment (asklar) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: Looks like this was merged with pending comments, @chrisglein can we update the bot to block merge until comments are resolved?


Filed #8650 to track the followup

Comment (chrisglein) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: > Looks like this was merged with pending comments, @chrisglein can we update the bot to block merge until comments are resolved?

If there are sign offs and no blocking reviews, today it's up to the review author to decide whether all conversations need to be resolved. If a reviewer has something that should block the PR, then they can use the blocking state.

The bot we use doesn't have the ability to enforce conversation state. It'd have to be branch policy. [There is one for that](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-conversation-resolution-before-merging). But we'd want to ensure everyone was on board with that.

Comment (vmoroz) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: @asklar, right, there is no extra copy: (1) creates the V8 value frm the provided buffer to access the script inside of V8, while the (2) returns the result by copying the compiled byte code from the V8 internal buffer into the result buffer.

Comment (JunielKatarn) [reply vnext/Shared/JSI/NapiJsiRuntime.cpp]: This comment missed my final sweep because it didn't show up in the files section after renaming the source file as `vnext/Microsoft.ReactNative.Cxx/JSI/NodeApiJsiRuntime.cpp`.
Will make sure to address all unresolved feedback in the future before setting `AutoMerge` :)
Note, this state of the code was already being used whenever we loaded V8, and is not a regression.

Thanks for creating the follow-up item. I'll assign to myself if it isn't already.

image

**Pull request merged**
